src/creature.cpp

```diff
@@ -324,24 +324,24 @@ void Creature::updateIcons() const
 {
 	SpectatorVec spectators;
 	g_game.map.getSpectators(spectators, position, true, true);
-	for (Creature* spectator : spectators) {
-		assert(dynamic_cast<Player*>(spectator) != nullptr);
-		static_cast<Player*>(spectator)->sendUpdateCreatureIcons(this);
+	for (const auto& spectator : spectators) {
+		assert(spectator->getPlayer() != nullptr);
+		std::static_pointer_cast<Player>(spectator)->sendUpdateCreatureIcons(getCreature());
 	}
 }
 
-void Creature::onCreatureAppear(Creature* creature, bool isLogin)
+void Creature::onCreatureAppear(std::shared_ptr<Creature> creature, bool isLogin)
 {
-	if (creature == this) {
+	if (creature.get() == this) {
```

Comment:
We’ve got some legacy code here, hehe!
```suggestion
	if (creature == getCreature()) {
```